File: 백준/Gold/1068. 트리/트리.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *


input = stdin.readline



n = int(input())

graph = [[] for _ in range(n)]

arr = list(map(int,input().split()))
k = int(input())

for i,parent in enumerate(arr):
    
    if parent == -1 or i == k: # 루트 이거나 제거할 노드 일 경우 그래프에 등록하지 않는다.
        continue
    
    graph[parent].append(i)

remove = []
remove.append(k)


while remove: # k와 연결된 모든 그래프는 -2(제거값)
    
    now = remove.pop()
    arr[now] = -2
    
    
    for nxt in graph[now]:
        if arr[nxt] != -2:
            remove.append(nxt)


ans = 0


for i in range(n):
    
    if arr[i] != -2 and not graph[i]: # 제거된 값이 아니고 자식이 없으면 , 리프 노드
        ans += 1

        
print(ans)

File: 백준/Gold/1068. 트리/트리.swift
import Foundation

let n = Int(readLine()!)!


var parents = readLine()!.split{ $0 == " " }.compactMap{ Int(String($0))! }

var children:[[Int]] = Array(repeating: [], count: n)

var target = Int(readLine()!)!

for (curr,parent)  in parents.enumerated() {
    
    if parent == -1 || curr == target {
        continue
    }
    
    children[parent].append(curr)
}



var remove:[Int] = []


remove.append(target)

while !remove.isEmpty {
    
    let now = remove.popLast()!
    
    parents[now] = -2 // 해당 노드 제거
    
    
    for child in children[now] { // 그 노드의 자식들을 지우기 위해 담는다.
        
        if parents[child] != -2 {  // 이미 제거된 노드가 아니고
            remove.append(child)
        }
        
        
    }
    
}


var ans:Int = 0

for i in  0..<n {
    
    if parents[i] != -2 && children[i].isEmpty { // 제거된 노드도 아니고, 자식이 없으면 = 리프 노드
        ans += 1
    }
    
}

print(ans)
